torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,186 @@
+#include <c10/util/irange.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/ir_views.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/frozen_concat_linear.h>
+#include <torch/csrc/jit/passes/frozen_conv_folding.h>
+#include <torch/csrc/jit/passes/frozen_graph_optimizations.h>
+#include <torch/csrc/jit/passes/remove_dropout.h>
+#include <torch/csrc/jit/passes/utils/optimization_pass.h>
+#include <torch/csrc/jit/runtime/graph_executor.h>
+#include <torch/csrc/utils/memory.h>
+#include <vector>
+#include "ATen/core/functional.h"
+#include "ATen/core/interned_strings.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+using Tensor = at::Tensor;
+
+class ConcatLinearLayers : public torch::jit::OptimizationPass {
+  using torch::jit::OptimizationPass::OptimizationPass;
+  void handleBlock(Block* b) override {
+    // We are using an ordered list so that one knows that
+    // one only needs to move arguments forward to be a valid move, not
+    // backwards.
+    std::unordered_map<Value*, std::list<Node*>> tensorUsers;
+
+    for (Node* n : b->nodes()) {
+      // Grouping together all linear layers that use the same Tensor for input
+      if (n->kind() != aten::linear) {
+        continue;
+      }
+      
+      // Punting on support for non-constant params for now.
```

Comment:
`for now` - isnt this somewhat permanent?

Reply:
It is somewhat fundamental in that if we don't have constant parameters we would have to concat at runtime, which has a performance penalty